### ISSUE TITLE
[Bug] IndexError attempting to list builds resulting in irrelevant error message

### ISSUE THREAD

**Issue by jaraco** [Azure/azure-devops-cli-extension#429]
**Describe the bug**

Attempting to list the builds for a pipeline, I bump into [an error](https://gist.github.com/c49922476a30c65028d4ef2aed1ec8c5).

**To Reproduce**
Azure Cli Version: 

```
$ az --version | grep azure-cli                                                                                                                                           
azure-cli (2.0.55)
```

Azure-Devops extension version: 

```
$ az --version | grep devops                                                                                                                                              
azure-devops (0.2.0)
```

1. login using az login
2. `az pipelines build list --organization jaraco --project pip-run`
3. See error

**Expected behavior**
Expected to see the list of builds for that pipeline or at least a better error message explaining what expectation was missed.

**Debug logs**

https://gist.github.com/ce354b016106a7d370f5e6d769aef7a9

**Additional context**
This is my first time attempting to use the az cli for Azure Devops, so I'm very new. I was using the vsts-cli earlier. [The project](https://dev.azure.com/jaraco/pip-run) is one I've just created today and is the only one in my org (which is just me).

I don't see any resources in the Azure portal. Is that expected?

**Comment by jaraco** [comment]
I also saw this error on login:

```
pip-run master $ az login                                                                                                                                                                
Note, we have launched a browser for you to login. For old experience with device code, use "az login --use-device-code"
You have logged in. Now let us find all the subscriptions to which you have access...
No subscriptions were found for 'None'. If this is expected, use '--allow-no-subscriptions' to have tenant level accesses
```

Perhaps that's related.

**Comment by atbagga** [comment]
You do not seem to have azure subscription associated with the account you used to login. 

As the message says.. try **az login --allow-no-subscriptions** 
and then try the devops commands.

**Comment by geverghe** [comment]
@jaraco - did that work for you?

**Comment by jaraco** [comment]
I tried logging in `--allow-no-subscriptions`, which suppresses the warning on login, but the az client still emits the same error as reported.

I was able to use the older vsts-cli, which was working... and of course I have pipelines set up (well [just one](https://dev.azure.com/jaraco/pip-run/)).

I thought perhaps the problem was related to _which_ account is associated with my pipelines. I have two Microsoft accounts associated with the same e-mail address (by necessity, not my choice), so I tried logging in with each. I think by default `az login` was using my "work" account (and has no subscriptions), so I tried a second time with my "personal" account. This time, `az login` showed two subscriptions:

```
$ az login                                                                                                                                                           
Note, we have launched a browser for you to login. For old experience with device code, use "az login --use-device-code"
You have logged in. Now let us find all the subscriptions to which you have access...
[
  {
    "cloudName": "AzureCloud",
    "id": "db9079df-1ed4-4475-a341-543afc87979d",
    "isDefault": true,
    "name": "Visual Studio Enterprise",
    "state": "Enabled",
    "tenantId": "39b48c3c-3b5d-42bd-bb10-7095bdc0749f",
    "user": {
      "name": "jaraco@jaraco.com",
      "type": "user"
    }
  },
  {
    "cloudName": "AzureCloud",
    "id": "0cd3e5fd-e192-4bed-9805-f9a744569d51",
    "isDefault": false,
    "name": "Free Trial",
    "state": "Disabled",
    "tenantId": "39b48c3c-3b5d-42bd-bb10-7095bdc0749f",
    "user": {
      "name": "jaraco@jaraco.com",
      "type": "user"
    }
  }
]
```

But still, `az pipelines ...` emits the IndexError.

**Comment by jaraco** [comment]
Can you help me understand the relationship between DevOps pipelines and Azure subscriptions? Is a subscription not needed? Is there a way to tell which Microsoft account my DevOps pipelines are associated with?

Okay - I've confirmed that my Azure DevOps is associated with my "work" address:

![image](https://user-images.githubusercontent.com/308610/52577010-391ac480-2def-11e9-8d2e-f2a25b266c06.png)

And that's the account that doesn't have any subscriptions.

**Comment by geverghe** [comment]
I see that in your trace the organization name is jarco. Could you give the fully qualified URL as organization and give it one more try (using the ID with which you were able to login, but it threw an error on az pipelines build list)?

`az pipelines build list --organization https://dev.azure.com/jarco --project pip-run`

**Comment by jaraco** [comment]
Okay, I seem to be getting closer:

```
$ az login --allow-no-subscriptions                                                                                                                                  
Note, we have launched a browser for you to login. For old experience with device code, use "az login --use-device-code"
You have logged in. Now let us find all the subscriptions to which you have access...
[
  {
    "cloudName": "AzureCloud",
    "id": "da4a5525-1805-4be8-913e-758c3b28915b",
    "isDefault": true,
    "name": "N/A(tenant level account)",
    "state": "Enabled",
    "tenantId": "da4a5525-1805-4be8-913e-758c3b28915b",
    "user": {
      "name": "jaraco@jaraco.com",
      "type": "user"
    }
  }
]
$ az pipelines build list --organization https://dev.azure.com/jaraco --project pip-run                                                                              
Before you can run Azure DevOps commands, you need to run the login command(az login if using AAD/MSA identity else az devops login if using PAT token) to setup credentials.  Please see https://aka.ms/azure-devops-cli-auth for more information.
```

I'll follow up on the hint. I suspect I need to use a PAT token, as that's what I was using for vsts-cli.

**Comment by geverghe** [comment]
Azure subscription is an azure concept and is used by Azure Pipelines when provisioning any resource on Azure. As per my understanding, if you are able to login and create a pipeline on Azure Pipelines using your work account which has no subscription, then you should be able to access the same pipelines via `azlogin` (using the --allow-no-subscriptions switch) and then running `az pipelines build list`. Let me know what happened after including the fully qualified URL for organization parameter in the list command.

**Comment by geverghe** [comment]
Are you running this on an installed client or cloudshell (shell.azure.com)? If on client, could you retry with `az logout` and `az login` and give it another go? If on cloudshell, just do `az login` one more time and it should start working for you. Let me know if this is not the case.

**Comment by jaraco** [comment]
After logging in with `az devops login`, the command succeeds (but only when the full URL is supplied):

```
$ az devops login                                                                                                                                                    
Token: 
$ az pipelines build list --organization https://dev.azure.com/jaraco --project pip-run                                                                              
[
  {
    "buildNumber": "20190207.3",
    "buildNumberRevision": 3,
...
```

I get the same IndexError if I don't supply the full URL.

I'm using an installed client (macOS Homebrew installed).

And I can confirm, after doing an `az logout` and `az login --allow-no-subscriptions`, the command is now succeeding:

```
~ $ az logout                                                                                                                                                                            
~ $ az login --allow-no-subscriptions                                                                                                                                                    
Note, we have launched a browser for you to login. For old experience with device code, use "az login --use-device-code"
You have logged in. Now let us find all the subscriptions to which you have access...
[
  {
    "cloudName": "AzureCloud",
    "id": "da4a5525-1805-4be8-913e-758c3b28915b",
    "isDefault": true,
    "name": "N/A(tenant level account)",
    "state": "Enabled",
    "tenantId": "da4a5525-1805-4be8-913e-758c3b28915b",
    "user": {
      "name": "jaraco@jaraco.com",
      "type": "user"
    }
  }
]
~ $ az pipelines build list --organization https://dev.azure.com/jaraco --project pip-run                                                                                                
[
  {
    "buildNumber": "20190207.3",
...
```

And to be sure, I did an `az devops logout` to ensure it wasn't _that_ credential that was allowing access, and I was still able to retrieve a result for `build list`.

To summarize, I think there are two issues at play here:

1. Because I have two Microsoft accounts under the same e-mail, it's easy to be confused as to which is the correct one to choose for a given service (I actually wonder if I should switch my organization over to the other ("personal") account, which has an MSDN subscription). Because of this, I was logging into the azure cli with the wrong account, one not associated with the DevOps org.
2. It wasn't clear from the error message that the `--organization` parameter was expecting the URL and not the name of the organization, leading to an obscure error message.

The latter issue is the main subject of my report. I still think there's something to be done here - specifically to provide a nicer error message if `--organization` doesn't look like a URL.

But based on the help, I believe I'm unblocked from further exploring use of the az cli as an interface for automating the enrollment of my projects with DevOps Pipelines. Thanks for the help.

**Comment by geverghe** [comment]
Thanks for the update @jaraco. Glad to hear that you are unblocked. When I tried it out, I was getting an error indicating that the fully qualified url was missing. Will try and repro this at our end and fix any ambiguous error statements. Keeping this open till we complete an investigation from our end.

**Comment by geverghe** [comment]
I am able to repro this issue. I was providing organization as myorg.visualstudio.com and it was throwing an error indicating it is an incorrect URL. However, on providing just myorg as organization value, it returns the same index out of range error. Need to improve the error message returned to indicate that full organization URL is expected.

**Comment by ishitam8** [comment]
This is the issue with how we parse URLs and not specific to build list. So it exists across all commands.

**Comment by atbagga** [comment]
Thank you for reporting the issue. 

Fixed in 0.11.0. 
Release ETA: Next week.
